test/modules/finance.spec.ts

```diff
@@ -313,17 +324,100 @@ describe('finance', () => {
       });
 
       describe('bitcoinAddress()', () => {
+        const m_legacy = /^1[A-HJ-NP-Za-km-z1-9]{25,39}$/;
+        const t_legacy = /^m[A-HJ-NP-Za-km-z1-9]{25,39}$/;
+        const m_segwit = /^3[A-HJ-NP-Za-km-z1-9]{25,39}$/;
+        const t_segwit = /^2[A-HJ-NP-Za-km-z1-9]{25,39}$/;
+        const m_bech32 = /^bc1[ac-hj-np-z02-9]{39,39}$/;
+        const t_bech32 = /^tb1[ac-hj-np-z02-9]{39,39}$/;
+        const m_taproot = /^bc1p[ac-hj-np-z02-9]{58,58}$/;
+        const t_taproot = /^tb1p[ac-hj-np-z02-9]{58,58}$/;
+
+        const isBtcAddress = (address: string) =>
+          [
+            m_legacy,
+            t_legacy,
+            m_segwit,
+            t_segwit,
+            m_bech32,
+            t_bech32,
+            m_taproot,
+            t_taproot,
+          ].some((r) => r.test(address));
+
         it('should return a valid bitcoin address', () => {
           const bitcoinAddress = faker.finance.bitcoinAddress();
-          /**
-           *  Note: Although the total length of a Bitcoin address can be 25-33 characters, regex quantifiers only check the preceding token
-           *  Therefore we take one from the total length of the address not including the first character ([13])
-           */
 
           expect(bitcoinAddress).toBeTruthy();
           expect(bitcoinAddress).toBeTypeOf('string');
-          expect(bitcoinAddress).toSatisfy(isValidBtcAddress);
-        });
+          expect(bitcoinAddress).toSatisfy(isBtcAddress);
+        });
+
+        it.each([
+          [BitcoinAddressFamily.Legacy, m_legacy],
+          [BitcoinAddressFamily.Segwit, m_segwit],
+          [BitcoinAddressFamily.Bech32, m_bech32],
+          [BitcoinAddressFamily.Taproot, m_taproot],
+        ])(
+          'should handle the type = $type argument',
+          (type: BitcoinAddressFamily, regex: RegExp) => {
```

Comment:
are these needed or can we use `as const` here?

```suggestion
        ] as const)(
          'should handle the type = $type argument',
          (type, regex) => {
```

Reply:
I'm curious about this change. It looks like asserting with `as const` and explicitly declaring the types produce the same effect of preventing entries with different types in the examples array. Is there any benefit on one vs the other or is it just readability ?

Reply:
> Is there any benefit on one vs the other or is it just readability ?

The explicit declaration of parameter types serves as the definitive source of truth, ensuring that functions receive the intended arguments. This approach emphasizes the developer's intent by stating, "This function **must** accept these specific argument types." Any discrepancies will be flagged as errors, such as "type string | number is not assignable to type string."

On the other hand, employing dynamic type inference through the use of "as const" makes the data as the primary source of truth. This conveys the message, "This data is established; functions should be capable of processing it accordingly." Errors will then be highlighted within the function's implementation, for instance, "Method 'toLowerCase' does not exist on type number," (you previously expected "string" but got "string | number") indicating an unanticipated data type.

Hope that helps. 

---

With all these information, I'd actually opt to not use "as const" in this case. While it is a great tool, it's not the right use case at this instance.

Reply:
I see the difference in the errors now, `as const` fails when we try to use the value in `faker.finance.bitcoinAddress`, not in the function declaration. It still does the job but I'm happy to revert. @Shinigami92  are you ok with that ?

Reply:
I might be trained a bit different, but for me: using types for parameters in a callback is mostly an anti-pattern and an indicator for wrongly typed library functions.
They can overshadow specific types and even **worse**, if they are miss-typed, the error is showing up at the _slightly_ wrong location ⚠️ 

good:
<img src="https://github.com/faker-js/faker/assets/7195563/8d4353c7-8e61-47df-a282-d200804e1fd4" width="300px" /> <img src="https://github.com/faker-js/faker/assets/7195563/1188ea66-8408-4050-90f9-8992c9a92d57" width="300px" />

bad:
<img src="https://github.com/faker-js/faker/assets/7195563/c3362914-ffb3-4505-ad85-cca4c97cd537" width="600px" />

Reply:
You use completely different types for your example. Of course ull receive errors in that case. Widen is a string union (literally enum) to a simple string is bound to fail...
Have to hardly disagree with your statement based on my previo8s provided statement.

Reply:
> using types for parameters in a callback is mostly an anti-pattern

Where did you get that from?

Reply:
> You use completely different types for your example. Of course you'll receive errors in that case. Widen is a string union (literally enum) to a simple string is bound to fail... I sadly have to disagree with your statement based on my previo8s provided statement.

exactly! that proves my point, and like I described: the error does not show up there or at the input where the values are provided but somewhere random downstream

> > using types for parameters in a callback is mostly an anti-pattern
> 
> Where did you get that from?

from my experience
reason described already above

Reply:
I'd like to add further clarity to our discussion by expanding on my previous comment.

> You use completely different types for your example. Of course you'll receive errors in that case. Widen is a string union (literally enum) to a simple string is bound to fail...

In my earlier message, I mentioned that such a mismatch would result in failure, but I didn't elaborate on why. The example encounters errors at "wrong locations" because it disregards the specified types for function parameters. As I've previously noted: Using static parameter type definitions guarantees that functions are supplied with the intended arguments. But in that case the function needs to be cabable to handly all the required parameters as well. This concept may be easier to illustrate by defining a function first and then use it's reference as a value.

Let's revisit your "bad" example:

```ts
function verifyTestCase(type: string, regex: Regexp | null) {
  const bitcoinAddress = faker.finance.bitcoinAddress({
    type, // fails
  });

  expect(bitcoinAddress).toBeTruthy();
  expect(bitcoinAddress).toBeTypeOf('string');
  expect(bitcoinAddress).toSatisfy(isBtcAddress);
  expect(bitcoinAddress).toMatch(regex); // fails
}

it.each(
  [BitcoinAddressFamily.Legacy, m_legacy],
  [BitcoinAddressFamily.Segwit, m_segwit],
  [BitcoinAddressFamily.Bech32, m_bech32],
  [BitcoinAddressFamily.Taproot, m_taproot],
)('will work but fail in function', verifyTestCase);
```

> the error does not show up there or at the input where the values are provided but somewhere random downstream

Yes, this example will fail for the same reasons as you’ve identified in your comment. Hoewver, it should demonstarte that these errors are justified, as the function is defined to potentiall receive nullish "regex" and a "type" argument that is of type "string". This emphasises the idea I stated, that defining static prameter types focuses on the developer's intend with the function's use case. 

I hope that this makes the intend of my previous comment more clear.

---

Let me clearify that I'm not saying that using dynamic type inference is wrong! I want to hint that both, dynamic and static type definitions, have their own use case which you, as the developer, should distinguish between. That being said, calling static type definitions an "anti-pattern", although `as const` was introduced to a much later in the TS ecosystem, is a bit of an overstatement I'd say. Furthermore, providing no evidence towards that thesis is quite unfair I'd say.

Reply:
you are right, my "anti-pattern"-call was only related to inline-anonymous-callbacks

Reply:
A good example for useful static parameter types can be found in [TypeScript's Beginner Guide](https://www.typescriptlang.org/docs/handbook/2/basic-types.html#explicit-types), where they encounter an upstream error after defining a function's parameter types. I have to admit that it's a really simple example, tho.

Reply:
> you are right, my "anti-pattern"-call was only related to inline-anonymous-callbacks

Good call. In that case I somewhat missunderstood you. Apologies for that.

Reply:
I checked our existing code and we use the `as const` a few times in our tests (although not overly often).

I also found one instance, where we use `(options, count: number)`. It is kind of hard to specifically search for these.

IMO the usage of `as const` makes sense in these cases, because the test signature isn't really the source of truth here.
You have the data, that you want to actually use in your tests, and they don't actually change their type just because you add types to the parameters. They only change how TS represents their type using the `as const` keyword.

I also found an example that requires the `as const` and does not work with parameter types:

- with parameter types

![grafik](https://github.com/faker-js/faker/assets/1579362/88ba506a-1f3d-4ed9-9dd7-381327b248bc)
![grafik](https://github.com/faker-js/faker/assets/1579362/7dbdc5e8-e301-4a6d-9c5e-5a745f73db5a)
![grafik](https://github.com/faker-js/faker/assets/1579362/d0947288-ba9a-4df6-ae98-c44484edbeba)

- with as const

![grafik](https://github.com/faker-js/faker/assets/1579362/b5173b15-72db-4bb2-946d-ece7cda515be)

I couldn't think of a way that would require the other way round for our tests, but that might just be observation bias.

A few times in our code we don't need either of `as const` or parameter types (and might even remove them).

I consider parameter types useful if they clarify the intend of the method/parameters or are otherwise intentionally placed to achieve a certain goal, however with the static data right above the method I think in this case the `as const` is sufficient and shorter.

Should we discuss this further in the next team meeting?

Reply:
> Should we discuss this further in the next team meeting?

As an optional bullet point if time is left over, sure 👍 

However I might ask Vitest-Team if they could think about generic-const 👀 like we did (`xy<const T>(...`)